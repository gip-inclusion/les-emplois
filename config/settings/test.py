import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
