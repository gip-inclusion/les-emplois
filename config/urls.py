from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path, register_converter
from django.views.generic import TemplateView

from itou.utils.urls import SiretConverter
from itou.www.dashboard import views as dashboard_views
from itou.www.login import views as login_views
from itou.www.signup import views as signup_views


register_converter(SiretConverter, "siret")

urlpatterns = [
    path("admin/", admin.site.urls),
    # --------------------------------------------------------------------------------------
    # allauth URLs. Order is important because some URLs are overriden.
    # --------------------------------------------------------------------------------------
    # Override allauth `account_signup` URL.
    # /accounts/signup/ <=> account_signup
    # We don't want any user to be able to signup using the default allauth `signup` url
    # because we have multiple specific signup processes for different kind of users.
    re_path(r"^accounts/signup/$", signup_views.signup),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_login` URL.
    # /accounts/login/ <=> account_login
    # We override this view because the login page should look slightly differently
    # for job seekers, prescribers and employers.
    # Also, PEAMU is only available for job seekers.
    re_path(r"^accounts/login/$", login_views.login),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_change_password` URL.
    # /accounts/password/change/ <=> account_change_password
    # https://github.com/pennersr/django-allauth/issues/468
    re_path(r"^accounts/password/change/$", dashboard_views.password_change),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_reset_password` URL.
    # Avoid user enumeration via password reset page.
    re_path(r"^accounts/password/reset/$", signup_views.ItouPasswordResetView.as_view()),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_logout` URL.
    # /accounts/logout/ <=> account_logout
    # We need custom code to process PEAMU logout.
    re_path(r"^accounts/logout/$", dashboard_views.logout),
    # --------------------------------------------------------------------------------------
    # Other allauth URLs.
    path("accounts/", include("allauth.urls")),
    # --------------------------------------------------------------------------------------
    # PEAMU URLs.
    path("accounts/", include("itou.allauth_adapters.peamu.urls")),
    # FranceConnect URLs.
    path("franceconnect/", include("itou.france_connect.urls")),
    # --------------------------------------------------------------------------------------
    # API.
    path("api/v1/", include("itou.api.urls", namespace="v1")),
    path("api/", include("itou.www.api.urls")),
    # www.
    path("", include("itou.www.home.urls")),
    path("apply/", include("itou.www.apply.urls")),
    path("approvals/", include("itou.www.approvals_views.urls")),
    path("autocomplete/", include("itou.www.autocomplete.urls")),
    path("dashboard/", include("itou.www.dashboard.urls")),
    path("employee_record/", include("itou.www.employee_record_views.urls")),
    path("institutions/", include("itou.www.institutions_views.urls")),
    path("invitations/", include("itou.www.invitations_views.urls")),
    path("prescribers/", include("itou.www.prescribers_views.urls")),
    path("search/", include("itou.www.search.urls")),
    path("siae/", include("itou.www.siaes_views.urls")),
    path("signup/", include("itou.www.signup.urls")),
    path("stats/", include("itou.www.stats.urls")),
    path("versions/", include("itou.www.releases.urls")),
    # Enable Anymailâ€™s status tracking
    # https://anymail.readthedocs.io/en/stable/esps/mailjet/#status-tracking-webhooks
    re_path(r"^webhooks/anymail/", include("anymail.urls")),
    path("welcoming_tour/", include("itou.www.welcoming_tour.urls")),
    # Static pages.
    path("accessibility/", TemplateView.as_view(template_name="static/accessibility.html"), name="accessibility"),
    # Errors pages.
    path("403/", TemplateView.as_view(template_name="403.html"), name="403"),
    path("404/", TemplateView.as_view(template_name="404.html"), name="404"),
    path("500/", TemplateView.as_view(template_name="500.html"), name="500"),
]

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    # In the local developpemnt environment, we enable the default callback URLs
    urlpatterns = [path("", include("itou.france_connect.urls"))] + urlpatterns
