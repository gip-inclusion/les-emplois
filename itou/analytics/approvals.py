from django.db.models import Count, Q

from itou.approvals import models as approvals_models
from itou.utils.apis import enums as api_enums

from . import models


def collect_approvals_counts(before):
    counts = approvals_models.Approval.objects.filter(created_at__lt=before).aggregate(
        total=Count("pk"),
        pe_notify_success=Count("pk", filter=Q(pe_notification_status=api_enums.PEApiNotificationStatus.SUCCESS)),
        pe_notify_pending=Count(
            "pk",
            filter=Q(
                pe_notification_status__in=(
                    api_enums.PEApiNotificationStatus.PENDING,
                    api_enums.PEApiNotificationStatus.SHOULD_RETRY,
                )
            ),
        ),
        pe_notify_error=Count("pk", filter=Q(pe_notification_status=api_enums.PEApiNotificationStatus.ERROR)),
    )
    return {
        models.DatumCode.APPROVAL_COUNT: counts["total"],
        models.DatumCode.APPROVAL_PE_NOTIFY_SUCCESS: counts["pe_notify_success"],
        models.DatumCode.APPROVAL_PE_NOTIFY_PENDING: counts["pe_notify_pending"],
        models.DatumCode.APPROVAL_PE_NOTIFY_ERROR: counts["pe_notify_error"],
    }


def collect_cancelled_approvals(before):
    cancelled_count = approvals_models.CancelledApproval.objects.filter(created_at__lt=before).count()
    return {models.DatumCode.APPROVAL_CANCELLED: cancelled_count}


def collect_analytics_data(before):
    return {
        **collect_approvals_counts(before),
        **collect_cancelled_approvals(before),
    }
