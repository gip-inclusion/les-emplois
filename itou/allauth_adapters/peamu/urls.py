from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
