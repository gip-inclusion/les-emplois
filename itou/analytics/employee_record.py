from django.db.models import Count, Max, Min

from itou.analytics import models
from itou.employee_record import models as employee_record_models


def collect_employee_records_count(before):
    return {
        models.DatumCode.EMPLOYEE_RECORD_COUNT: employee_record_models.EmployeeRecord.objects.filter(
            created_at__lt=before
        ).count(),
    }


def collect_probably_deleted_employee_records(before):
    data = employee_record_models.EmployeeRecord.objects.filter(created_at__lt=before).aggregate(
        first_pk=Min("pk"),
        last_pk=Max("pk"),
        count_pk=Count("pk"),
    )

    value = ((data["last_pk"] - data["first_pk"] + 1) - data["count_pk"]) if None not in data.values() else 0
    return {models.DatumCode.EMPLOYEE_RECORD_DELETED: value}


def collect_employee_records_processing_code_of_first_exchange(before):
    data = (
        employee_record_models.EmployeeRecord.objects.filter(created_at__lt=before)
        .values("asp_processing_code")
        .annotate(
            count=Count("asp_processing_code"),
        )
    )
    count_by_processing_code = {item["asp_processing_code"]: item["count"] for item in data}

    return {
        models.DatumCode.EMPLOYEE_RECORD_PROCESSED_AT_FIRST_EXCHANGE: count_by_processing_code.get(
            employee_record_models.EmployeeRecord.ASP_PROCESSING_SUCCESS_CODE, 0
        ),
        models.DatumCode.EMPLOYEE_RECORD_WITH_ERROR_AT_FIRST_EXCHANGE: sum(
            count
            for code, count in count_by_processing_code.items()
            if code != employee_record_models.EmployeeRecord.ASP_PROCESSING_SUCCESS_CODE
        ),
        models.DatumCode.EMPLOYEE_RECORD_WITH_ERROR_3436_AT_FIRST_EXCHANGE: count_by_processing_code.get(
            employee_record_models.EmployeeRecord.ASP_DUPLICATE_ERROR_CODE, 0
        ),
    }


def collect_employee_record_with_at_least_one_error(before):
    employee_records_with_at_least_one_error = employee_record_models.EmployeeRecordQuerySet.union(
        employee_record_models.EmployeeRecord.objects.filter(created_at__lt=before)
        .exclude(
            asp_processing_code=employee_record_models.EmployeeRecord.ASP_PROCESSING_SUCCESS_CODE,
        )
        .values("pk"),
        employee_record_models.EmployeeRecordUpdateNotification.objects.filter(created_at__lt=before)
        .exclude(
            asp_processing_code=employee_record_models.EmployeeRecord.ASP_PROCESSING_SUCCESS_CODE,
        )
        .values("employee_record"),
    )

    value = (
        employee_record_models.EmployeeRecord.objects.filter(pk__in=employee_records_with_at_least_one_error)
        .distinct()
        .count()
    )
    return {models.DatumCode.EMPLOYEE_RECORD_WITH_AT_LEAST_ONE_ERROR: value}


def collect_analytics_data(before):
    return {
        **collect_employee_records_count(before),
        **collect_probably_deleted_employee_records(before),
        **collect_employee_records_processing_code_of_first_exchange(before),
        **collect_employee_record_with_at_least_one_error(before),
    }
