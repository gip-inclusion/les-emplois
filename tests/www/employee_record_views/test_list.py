import datetime

import pytest
from dateutil.relativedelta import relativedelta
from django.test import override_settings
from django.urls import reverse
from django.utils import timezone

from itou.employee_record.enums import Status
from itou.users.enums import LackOfNIRReason
from itou.utils.templatetags import format_filters
from tests.approvals import factories as approvals_factories
from tests.companies.factories import CompanyWithMembershipAndJobsFactory
from tests.employee_record import factories as employee_record_factories
from tests.employee_record.factories import EmployeeRecordFactory
from tests.job_applications.factories import JobApplicationWithApprovalNotCancellableFactory
from tests.utils.test import BASE_NUM_QUERIES, TestCase, assertMessages, parse_response_to_soup


@pytest.mark.usefixtures("unittest_compatibility")
class ListEmployeeRecordsTest(TestCase):
    def setUp(self):
        super().setUp()
        # User must be super user for UI first part (tmp)
        self.company = CompanyWithMembershipAndJobsFactory(name="Evil Corp.", membership__user__first_name="Elliot")
        self.company_without_perms = CompanyWithMembershipAndJobsFactory(
            kind="EITI", name="A-Team", membership__user__first_name="Hannibal"
        )
        self.user = self.company.members.get(first_name="Elliot")
        self.user_without_perms = self.company_without_perms.members.get(first_name="Hannibal")
        self.job_application = JobApplicationWithApprovalNotCancellableFactory(to_company=self.company)
        self.job_seeker = self.job_application.job_seeker
        self.url = reverse("employee_record_views:list")

    def test_permissions(self):
        """
        Non-eligible SIAE should not be able to access this list
        """
        self.client.force_login(self.user_without_perms)

        response = self.client.get(self.url)

        assert response.status_code == 403

    def test_new_employee_records(self):
        """
        Check if new employee records / job applications are displayed in the list
        """
        self.client.force_login(self.user)

        response = self.client.get(self.url)

        self.assertContains(response, format_filters.format_approval_number(self.job_application.approval.number))

    def test_status_filter(self):
        """
        Check status filter
        """
        # No status defined
        self.client.force_login(self.user)
        approval_number_formatted = format_filters.format_approval_number(self.job_application.approval.number)

        response = self.client.get(self.url)
        self.assertContains(response, approval_number_formatted)

        # Or NEW
        response = self.client.get(self.url + "?status=NEW")
        self.assertContains(response, approval_number_formatted)

        # More complete tests to come with fixtures files
        for status in [Status.SENT, Status.REJECTED, Status.PROCESSED]:
            response = self.client.get(self.url + f"?status={status.value}")
            self.assertNotContains(response, approval_number_formatted)

    def test_job_seeker_filter(self):
        approval_number_formatted = format_filters.format_approval_number(self.job_application.approval.number)
        other_job_application = JobApplicationWithApprovalNotCancellableFactory(to_company=self.company)
        other_approval_number_formatted = format_filters.format_approval_number(other_job_application.approval.number)
        self.client.force_login(self.user)

        response = self.client.get(self.url)
        self.assertContains(response, approval_number_formatted)
        self.assertContains(response, other_approval_number_formatted)

        response = self.client.get(self.url + f"?job_seekers={self.job_seeker.pk}")
        self.assertContains(response, approval_number_formatted)
        self.assertNotContains(response, other_approval_number_formatted)

        response = self.client.get(self.url + "?job_seekers=0")
        self.assertContains(response, "Sélectionnez un choix valide. 0 n’en fait pas partie.")
        self.assertContains(response, approval_number_formatted)
        self.assertContains(response, other_approval_number_formatted)

    def test_employee_records_approval_display(self):
        self.client.force_login(self.user)
        approval = self.job_application.approval
        approval.start_at = datetime.date(2023, 9, 2)
        approval.end_at = datetime.date(2024, 10, 11)
        approval.save()

        response = self.client.get(self.url)

        self.assertContains(response, "Date de début : <b>02/09/2023</b>")
        self.assertContains(response, "Date de fin prévisionnelle : <b>11/10/2024</b>")

    def test_employee_records_with_a_suspension_need_to_be_updated(self):
        self.client.force_login(self.user)
        approvals_factories.SuspensionFactory(
            approval=self.job_application.approval, siae=self.job_application.to_company
        )

        response = self.client.get(self.url + "?status=NEW")

        # Global message alert
        assert str(parse_response_to_soup(response, selector=".s-title-01 .alert")) == self.snapshot(name="alert")

        # Item message alert
        assert str(
            parse_response_to_soup(
                response,
                selector=".employee-records-list .card-body > div:last-child",
                replace_objects_pk_in_href=[self.job_application],
            )
        ) == self.snapshot(name="action")

    def test_existing_employee_records_with_a_suspension_does_not_show_need_to_be_updated_message(self):
        self.client.force_login(self.user)
        employee_record = EmployeeRecordFactory(job_application=self.job_application)
        approvals_factories.SuspensionFactory(
            approval=self.job_application.approval, siae=self.job_application.to_company
        )

        response = self.client.get(self.url + "?status=NEW")

        # Global message alert
        assertMessages(response, [])

        # Item message alert
        assert (
            str(
                parse_response_to_soup(
                    response,
                    selector=".employee-records-list .card-body > div:last-child",
                    replace_objects_pk_in_href=[self.job_application, employee_record],
                )
            )
            == self.snapshot()
        )

    def test_employee_records_with_a_prolongation_need_to_be_updated(self):
        self.client.force_login(self.user)
        approvals_factories.ProlongationFactory(
            approval=self.job_application.approval,
            declared_by_siae=self.job_application.to_company,
        )

        response = self.client.get(self.url + "?status=NEW")

        # Global message alert
        assert str(parse_response_to_soup(response, selector=".s-title-01 .alert")) == self.snapshot(name="alert")
        # Item message alert
        assert str(
            parse_response_to_soup(
                response,
                selector=".employee-records-list .card-body > div:last-child",
                replace_objects_pk_in_href=[self.job_application],
            )
        ) == self.snapshot(name="action")

    def test_existing_employee_records_with_a_prolongation_does_not_show_need_to_be_updated_message(self):
        self.client.force_login(self.user)
        employee_record = EmployeeRecordFactory(job_application=self.job_application)
        approvals_factories.ProlongationFactory(
            approval=self.job_application.approval,
            declared_by_siae=self.job_application.to_company,
        )

        response = self.client.get(self.url + "?status=NEW")

        # Global message alert
        assertMessages(response, [])
        # Item message alert
        assert (
            str(
                parse_response_to_soup(
                    response,
                    selector=".employee-records-list .card-body > div:last-child",
                    replace_objects_pk_in_href=[self.job_application, employee_record],
                )
            )
            == self.snapshot()
        )

    def test_employee_record_to_disable(self):
        self.client.force_login(self.user)
        employee_record = employee_record_factories.EmployeeRecordFactory(job_application=self.job_application)

        response = self.client.get(self.url + "?status=NEW")

        assert (
            str(
                parse_response_to_soup(
                    response,
                    selector=".employee-records-list .card-body > div:last-child",
                    replace_objects_pk_in_href=[self.job_application, employee_record],
                )
            )
            == self.snapshot()
        )

    @override_settings(TALLY_URL="https://tally.so")
    def test_employee_records_with_nir_associated_to_other(self):
        self.client.force_login(self.user)
        self.job_seeker.nir = ""
        self.job_seeker.lack_of_nir_reason = LackOfNIRReason.NIR_ASSOCIATED_TO_OTHER
        self.job_seeker.save(update_fields=("nir", "lack_of_nir_reason"))

        response = self.client.get(self.url + "?status=NEW")

        self.assertContains(response, format_filters.format_approval_number(self.job_application.approval.number))
        # Global message alert
        assert str(parse_response_to_soup(response, selector=".s-title-01 .alert")) == self.snapshot(name="alert")
        # Item message alert
        assert str(
            parse_response_to_soup(
                response,
                selector=".employee-records-list .card-body > div:last-child",
                replace_objects_pk_in_href=[self.job_application],
            )
        ) == self.snapshot(name="action")

    @override_settings(TALLY_URL="https://tally.so")
    def test_employee_record_to_disable_with_nir_associated_to_other(self):
        self.client.force_login(self.user)
        self.job_seeker.nir = ""
        self.job_seeker.lack_of_nir_reason = LackOfNIRReason.NIR_ASSOCIATED_TO_OTHER
        self.job_seeker.save(update_fields=("nir", "lack_of_nir_reason"))
        new_er = employee_record_factories.EmployeeRecordFactory(job_application=self.job_application)

        response = self.client.get(self.url + "?status=NEW")

        self.assertContains(response, format_filters.format_approval_number(self.job_application.approval.number))
        # Global message alert
        assert str(parse_response_to_soup(response, selector=".s-title-01 .alert")) == self.snapshot(name="alert")
        # Item message alert
        assert str(
            parse_response_to_soup(
                response,
                selector=".employee-records-list .card-body > div:last-child",
                replace_objects_pk_in_href=[new_er],
            )
        ) == self.snapshot(name="action")

    def test_rejected_without_custom_message(self):
        self.client.force_login(self.user)

        record = employee_record_factories.EmployeeRecordWithProfileFactory(job_application__to_company=self.company)
        record.update_as_ready()
        record.update_as_sent(self.faker.asp_batch_filename(), 1, None)
        record.update_as_rejected("0012", "JSON Invalide", None)

        response = self.client.get(self.url + "?status=REJECTED")
        self.assertContains(response, "Erreur 0012")
        self.assertContains(response, "JSON Invalide")

    def test_rejected_custom_messages(self):
        self.client.force_login(self.user)

        record = employee_record_factories.EmployeeRecordWithProfileFactory(job_application__to_company=self.company)

        tests_specs = [
            (
                "3308",
                "Le champ Commune de Naissance doit être en cohérence avec le champ Département de Naissance",
                "Il semblerait que la commune de naissance sélectionnée ne corresponde pas au département",
            ),
            (
                "3417",
                "Le code INSEE de la commune de l’adresse doit correspondre à un code INSEE de commune référencée",
                "La commune de résidence du salarié n’est pas référencée",
            ),
            (
                "3435",
                "L’annexe de la structure doit être à l’état Valide ou Provisoire",
                "Nous n’avons pas encore reçu d’annexe financière à jour pour votre structure.",
            ),
            (
                "3436",
                "Un PASS IAE doit être unique pour un même SIRET",
                "La fiche salarié associée à ce PASS IAE et à votre SIRET a déjà été intégrée à l’ASP.",
            ),
        ]
        for err_code, err_message, custom_err_message in tests_specs:
            with self.subTest(err_code):
                record.status = Status.SENT
                record.update_as_rejected(err_code, err_message, "{}")

                response = self.client.get(self.url + "?status=REJECTED")
                self.assertContains(response, f"Erreur {err_code}")
                self.assertNotContains(response, err_message)
                self.assertContains(response, custom_err_message)

    def _check_employee_record_order(self, url, first_job_application, second_job_application):
        response = self.client.get(url)
        response_text = response.content.decode(response.charset)
        # The index method raises ValueError if the value isn't found
        first_job_seeker_position = response_text.index(
            format_filters.format_approval_number(first_job_application.approval.number)
        )
        second_job_seeker_position = response_text.index(
            format_filters.format_approval_number(second_job_application.approval.number)
        )
        assert first_job_seeker_position < second_job_seeker_position

    def test_new_employee_records_sorted(self):
        """
        Check if new employee records / job applications are correctly sorted
        """
        self.client.force_login(self.user)

        job_applicationA = JobApplicationWithApprovalNotCancellableFactory(
            to_company=self.company,
            job_seeker__last_name="Aaaaa",
            hiring_start_at=timezone.now() - relativedelta(days=15),
        )
        job_applicationZ = JobApplicationWithApprovalNotCancellableFactory(
            to_company=self.company,
            job_seeker__last_name="Zzzzz",
            hiring_start_at=timezone.now() - relativedelta(days=10),
        )

        # Zzzzz's hiring start is more recent
        self._check_employee_record_order(self.url, job_applicationZ, job_applicationA)

        # order with -hiring_start_at is the default
        self._check_employee_record_order(self.url + "?order=-hiring_start_at", job_applicationZ, job_applicationA)
        self._check_employee_record_order(self.url + "?order=hiring_start_at", job_applicationA, job_applicationZ)

        # Zzzzz after Aaaaa
        self._check_employee_record_order(self.url + "?order=name", job_applicationA, job_applicationZ)
        self._check_employee_record_order(self.url + "?order=-name", job_applicationZ, job_applicationA)

        # Count queries
        num_queries = BASE_NUM_QUERIES
        num_queries += 1  # Get django session
        num_queries += 3  # Get current user and siae
        num_queries += 1  # Select job seeker for filters
        num_queries += 1  # Select employee_records status count
        num_queries += 1  # Select ordered job applications
        num_queries += 1  # Select EmployeeRecords
        with self.assertNumQueries(num_queries):
            self.client.get(self.url)

    def test_rejected_employee_records_sorted(self):
        self.client.force_login(self.user)

        recordA = employee_record_factories.EmployeeRecordWithProfileFactory(
            job_application__to_company=self.company,
            job_application__job_seeker__last_name="Aaaaa",
            job_application__hiring_start_at=timezone.now() - relativedelta(days=15),
        )
        recordZ = employee_record_factories.EmployeeRecordWithProfileFactory(
            job_application__to_company=self.company,
            job_application__job_seeker__last_name="Zzzzz",
            job_application__hiring_start_at=timezone.now() - relativedelta(days=10),
        )
        for i, record in enumerate((recordA, recordZ)):
            record.update_as_ready()
            record.update_as_sent(f"RIAE_FS_2021041013000{i}.json", 1, None)
            record.update_as_rejected("0012", "JSON Invalide", None)

        # Zzzzz's hiring start is more recent
        self._check_employee_record_order(
            self.url + "?status=REJECTED", recordZ.job_application, recordA.job_application
        )

        # order with -hiring_start_at is the default
        self._check_employee_record_order(
            self.url + "?status=REJECTED&order=-hiring_start_at",
            recordZ.job_application,
            recordA.job_application,
        )
        self._check_employee_record_order(
            self.url + "?status=REJECTED&order=hiring_start_at",
            recordA.job_application,
            recordZ.job_application,
        )

        # Zzzzz after Aaaaa
        self._check_employee_record_order(
            self.url + "?status=REJECTED&order=name",
            recordA.job_application,
            recordZ.job_application,
        )
        self._check_employee_record_order(
            self.url + "?status=REJECTED&order=-name",
            recordZ.job_application,
            recordA.job_application,
        )

    def test_ready_employee_records_sorted(self):
        self.client.force_login(self.user)

        recordA = employee_record_factories.EmployeeRecordWithProfileFactory(
            job_application__to_company=self.company,
            job_application__job_seeker__last_name="Aaaaa",
            job_application__hiring_start_at=timezone.now() - relativedelta(days=15),
        )
        recordZ = employee_record_factories.EmployeeRecordWithProfileFactory(
            job_application__to_company=self.company,
            job_application__job_seeker__last_name="Zzzzz",
            job_application__hiring_start_at=timezone.now() - relativedelta(days=10),
        )
        for record in (recordA, recordZ):
            record.update_as_ready()

        # Zzzzz's hiring start is more recent
        self._check_employee_record_order(self.url + "?status=READY", recordZ.job_application, recordA.job_application)

        # order with -hiring_start_at is the default
        self._check_employee_record_order(
            self.url + "?status=READY&order=-hiring_start_at",
            recordZ.job_application,
            recordA.job_application,
        )
        self._check_employee_record_order(
            self.url + "?status=READY&order=hiring_start_at",
            recordA.job_application,
            recordZ.job_application,
        )

        # Zzzzz after Aaaaa
        self._check_employee_record_order(
            self.url + "?status=READY&order=name",
            recordA.job_application,
            recordZ.job_application,
        )
        self._check_employee_record_order(
            self.url + "?status=READY&order=-name",
            recordZ.job_application,
            recordA.job_application,
        )

    def test_display_result_count(self):
        self.client.force_login(self.user)
        response = self.client.get(self.url + "?status=NEW")
        self.assertContains(response, "1 résultat")

        JobApplicationWithApprovalNotCancellableFactory(to_company=self.company)
        response = self.client.get(self.url + "?status=NEW")
        self.assertContains(response, "2 résultats")

        response = self.client.get(self.url + "?status=READY")
        self.assertContains(response, "0 résultat")
