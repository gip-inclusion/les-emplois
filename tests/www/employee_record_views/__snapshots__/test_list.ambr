# serializer version: 1
# name: ListEmployeeRecordsTest.test_employee_record_to_disable
  '''
  <div class="c-box--results__footer">
                  <div class="d-flex flex-column flex-md-row justify-content-md-end gap-3">
                      
                          <a aria-label="Désactiver la fiche salarié de John DOE" class="btn btn-outline-primary btn-block w-100 w-md-auto" href="/employee_record/disable/[PK of EmployeeRecord]">Désactiver</a>
                      
                      <a aria-label="Compléter la fiche salarié de John DOE" class="btn btn-primary btn-block w-100 w-md-auto" href="/employee_record/create/[PK of JobApplication]?from_status=NEW">Compléter</a>
                  </div>
              </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_record_to_disable_with_nir_associated_to_other[action]
  '''
  <div class="c-box--results__footer">
                  <a aria-label="Régulariser le n° de sécurité sociale John DOE (ouverture dans un nouvel onglet)" class="btn btn-primary btn-block w-100 w-md-auto has-external-link" href="https://tally.so/r/wzxQlg?employeerecord=[PK of EmployeeRecord]" rel="noopener" target="_blank">Régulariser le n° de sécurité sociale</a>
              </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_record_to_disable_with_nir_associated_to_other[alert]
  '''
  <div class="alert alert-warning">
              <div class="row">
                  <div class="col-auto pe-0">
                      <i class="ri-error-warning-line ri-xl text-danger"></i>
                  </div>
                  <div class="col">
                      <p class="mb-0">
                          <strong>Une action de votre part est nécessaire</strong>
                      </p>
                      <p class="mb-0">
                          Attention, nous avons détecté une ou plusieurs fiches salarié qui nécessitent une régularisation manuelle de votre part.
                      </p>
                  </div>
              </div>
          </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_records_with_a_prolongation_need_to_be_updated[action]
  '''
  <div class="c-box--results__footer">
                  <a aria-label="Mettre à jour la fiche salarié de John DOE" class="btn btn-primary btn-block w-100 w-md-auto" href="/employee_record/create/[PK of JobApplication]">Mettre à jour</a>
              </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_records_with_a_prolongation_need_to_be_updated[alert]
  '''
  <div class="alert alert-warning">
              <div class="row">
                  <div class="col-auto pe-0">
                      <i class="ri-error-warning-line ri-xl text-danger"></i>
                  </div>
                  <div class="col">
                      <p class="mb-0">
                          <strong>Une action de votre part est nécessaire</strong>
                      </p>
                      <p class="mb-0">
                          Attention, nous avons détecté une ou plusieurs fiches salarié qui nécessitent une régularisation manuelle de votre part.
                      </p>
                  </div>
              </div>
          </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_records_with_a_suspension_need_to_be_updated[action]
  '''
  <div class="c-box--results__footer">
                  <a aria-label="Mettre à jour la fiche salarié de John DOE" class="btn btn-primary btn-block w-100 w-md-auto" href="/employee_record/create/[PK of JobApplication]">Mettre à jour</a>
              </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_records_with_a_suspension_need_to_be_updated[alert]
  '''
  <div class="alert alert-warning">
              <div class="row">
                  <div class="col-auto pe-0">
                      <i class="ri-error-warning-line ri-xl text-danger"></i>
                  </div>
                  <div class="col">
                      <p class="mb-0">
                          <strong>Une action de votre part est nécessaire</strong>
                      </p>
                      <p class="mb-0">
                          Attention, nous avons détecté une ou plusieurs fiches salarié qui nécessitent une régularisation manuelle de votre part.
                      </p>
                  </div>
              </div>
          </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_records_with_nir_associated_to_other[action]
  '''
  <div class="c-box--results__footer">
                  <a aria-label="Régulariser le n° de sécurité sociale John DOE (ouverture dans un nouvel onglet)" class="btn btn-primary btn-block w-100 w-md-auto has-external-link" href="https://tally.so/r/wzxQlg?jobapplication=[PK of JobApplication]" rel="noopener" target="_blank">Régulariser le n° de sécurité sociale</a>
              </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_employee_records_with_nir_associated_to_other[alert]
  '''
  <div class="alert alert-warning">
              <div class="row">
                  <div class="col-auto pe-0">
                      <i class="ri-error-warning-line ri-xl text-danger"></i>
                  </div>
                  <div class="col">
                      <p class="mb-0">
                          <strong>Une action de votre part est nécessaire</strong>
                      </p>
                      <p class="mb-0">
                          Attention, nous avons détecté une ou plusieurs fiches salarié qui nécessitent une régularisation manuelle de votre part.
                      </p>
                  </div>
              </div>
          </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_existing_employee_records_with_a_prolongation_does_not_show_need_to_be_updated_message
  '''
  <div class="c-box--results__footer">
                  <div class="d-flex flex-column flex-md-row justify-content-md-end gap-3">
                      
                          <a aria-label="Désactiver la fiche salarié de John DOE" class="btn btn-outline-primary btn-block w-100 w-md-auto" href="/employee_record/disable/[PK of EmployeeRecord]">Désactiver</a>
                      
                      <a aria-label="Compléter la fiche salarié de John DOE" class="btn btn-primary btn-block w-100 w-md-auto" href="/employee_record/create/[PK of JobApplication]?from_status=NEW">Compléter</a>
                  </div>
              </div>
  '''
# ---
# name: ListEmployeeRecordsTest.test_existing_employee_records_with_a_suspension_does_not_show_need_to_be_updated_message
  '''
  <div class="c-box--results__footer">
                  <div class="d-flex flex-column flex-md-row justify-content-md-end gap-3">
                      
                          <a aria-label="Désactiver la fiche salarié de John DOE" class="btn btn-outline-primary btn-block w-100 w-md-auto" href="/employee_record/disable/[PK of EmployeeRecord]">Désactiver</a>
                      
                      <a aria-label="Compléter la fiche salarié de John DOE" class="btn btn-primary btn-block w-100 w-md-auto" href="/employee_record/create/[PK of JobApplication]?from_status=NEW">Compléter</a>
                  </div>
              </div>
  '''
# ---
